---
title: 'Information-Content-Informed Kendall-tau Correlation: Utilizing Missing Values'
author: 'Robert M Flight & Hunter NB Moseley'
date: '`r Sys.time()`'
output: 
  rmarkdown::word_document
bibliography: '`r here::here("doc/icikt_references.json")`'
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE, fig.wdith = 8, fig.height = 5)
```

## Abstract

Almost all correlation measures currently available are unable to handle missing values.
Typically, missing values are either ignored completely by removing them or are imputed and used in the calculation of the correlation coefficient.
In both cases, the correlation value will be impacted based on a perspective that missing data represents no useful information.
However, missing values occur in real data sets for a variety of reasons. 
In -omics data sets that are derived from analytical measurements, the primary reason is that a specific measureable phenomenom falls below the detection limits of the analytical instrumentation.
Therefore, these missing data are not missing at random, but represent some information by their "missingness".
Therefore, we propose an information-content-informed Kendall-tau (ICI-Kt) correlation coefficient that allows missing values to carry explicit information in determination of concordant and discordant pairs.
With both simulated and real data sets from RNA-seq experiments, we demonstrate that the ICI-Kt allows for the inclusion of missing data values as interpretable information.
Moreover, our implementation of ICI-Kt uses a mergesort-like algorithm that provides nlog(n) computational performance. 
Finally, we show that approximate ICI-Kt values can be calculated using smaller subsets of large data sets with significant time savings, which has practical computational value when feature sizes are very large.

The ICI-Kt correlation calculation is available in an R package on GitHub at https://moseleyBionformaticsLab/ICIKendallTau.

## Introduction

Correlation as a measure of the relatedness or similarity of two or more sets of data has a long history, with the mathematical technique being used (and abused) in various scientific fields since it's introduction [@pearson_notes_1920; @rodgers_13_1988].
More recently, correlation calculations have become a cornerstone statistical method in the analysis and integration of varied -omics' datasets, especially the big 5 omics: genomics, transcriptomics, proteomics, metabolomics, epigenomics [@gu_complexheatmap_2016].
Correlation between biomolecular features (nucleotide variants, RNA transcripts, proteins, metabolites) may be used to determine the relationship between pairs of the features as well as to detect and derive correlative structures between groups of features [@fukushima_integratedomics_2009].
Moreover, correlation is a foundational method for generating biomolecular feature - feature interaction networks, like those provided by STRING [@szklarczyk_string_2017], Genemania [@franz_genemania_2018], and WCGNA [@langfelder_wgcna_2008].
Feature - feature correlation may also be used to inform which features are used for imputation of missing values [@faquih_missingvalueworkflow_2020].
Correlation of samples in -omics may be used for evaluating whether all the samples from a particular condition are truly related to each other, as well as to inform the choice of analysis, and to check for outliers prior to other statistical analysis [@flight_timecourseexploration_2010; @gierlinski_statisticalmodels_2015].

However, all analytical methods, and in particular the analytical methods used in -omics where so many analytes are being measured simultaneously; suffer from missing measurements.
Of course some analytes will be missing at random just because of some issue with the instrument or that particular sample, but a large amount of missing measurements are left-censored due to analytes being below the effective detection limit of the instrument, as shown in Figure 1.
Some analytical instruments are actually purposely designed to floor measurements, when they occur below a certain signal to noise ratio threshold.
Imputation of missing measurements in -omics samples is an active area of research, which we will not cover here beyond to say that it is worthwhile and very necessary in many instances.
When it comes to calculating correlation, there are very few methods that explicitly account for missing data that we know of.
In many cases, missing values may be imputed to zero (or another value) and then included in the correlation calculation.
Imputation or replacement with zero is likely to lead to high inflation of Pearson and related correlation measures.
Other approaches keep only those features with no missing values across samples (complete cases), or keep non-missing between any pair of samples (pairwise complete cases).
Each of these are likely to mis-estimate the real sample - sample correlation values, especially from specific data interpretation perspectives.

```{r include_figure1}
knitr::include_graphics(here::here("doc/the_problem.png"))
```

Figure 1. Graphical description of the left-censored data problem.
The true analyte concentration range covers the full range of the density distribution, with the minimum on the left, and the maximum on the right.
An example density plot of the analyte concentrations for a single sample is shown as a solid line, and the instrument response to concentration as a dashed line.
Below certain analyte concentrations, highlighted as the red region, the instrument returns **0** or **missing** values.
Above certain concentrations, highlighted as the yellow region, all values returned will also be identical (or flagged depending on the instrument).
Which analytes will have concentrations in the red region may vary from sample to sample due to the overall sample composition, as well as natural variances within each sample.

Assuming that a majority of missing values are **not missing at random**, but rather result from left-censored distributions due to the analyte being below the effective detection limit (see Figure 1), we propose that these missing values do in fact encode useful information that can be incorporated into correlation calculations, especially the Kendall-tau paired rank correlation coefficient.
In this manuscript, we propose new definitions of concordant and discordant rank pairs that include missing values, as well as methods for incorporating missing values into the number of tied values for the equivalent of the modified Kendall-tau coefficient.
We examine the effect of missing values on simulated data, comparing our information-content-informed Kendall-tau (ICI-Kt) method with other simpler methods of handling the missing values, namely removing them or imputing them to zero.
Finally, we examine the ability to recapitulate ICI-Kt correlation values calculated on a large feature -omics data set using feature subsets, which is useful as the run time of very large feature data sets can become prohibitive.

## Methods

### Definition of Concordant and Discordant Pairs

In the simplest form, the Kendall-tau correlation can be defined as:

$$\tau = \frac{ | pairs_{concordant}  | - | pairs_{discordant}  |}{\left | pairs_{concordant} \right | + \left | pairs_{discordant} \right |}$$
**NEED to discuss this definition with Hunter**.In this case a **pair** are any two $x$ and $y$ points from two different samples, consisting of $(x_i, x_j)$ and $(y_i, y_j)$.
A concordant pair has the following classical definition:

  * $x_i > x_j$ and $y_i > y_j$ 
  * $x_i < x_j$ and $y_i < y_j$

A discordant pair has the following classical definition [@kendall_newmeasure_1938]:

  * $x_i > x_j$ and $y_i < y_j$
  * $x_i < x_j$ and $y_i > y_j$

Assuming that the majority of missing values are not missing at random, but due to being below the effective detection limit, we can expand the concordant and discordant pair definitions to include missing values. Here $!x$ indicates a missing value.
The information-content-informed concordant pair definitions are below:

  * $x_i > x_j$ and $y_i > y_j$
  * $x_i < x_j$ and $y_i < y_j$
  * $x_i > x_j$ and $y_i \& !y_j$
  * $x_i < x_j$ and $!y_i \& y_j$
  * $x_i \& !x_j$ and $y_i > y_j$
  * $!x_i \& x_j$ and $y_i < y_j$ 
  * $x_i \& !x_j$ and $y_i \& !y_j$ 
  * $x_i \& x_j$ and $!y_i \& y_j$ 
  
The information content informed discordant pair definitions are below:

  * $x_i > x_j$ and $y_i < y_j$ 
  * $x_i < x_j$ and $y_i > y_j$
  * $x_i > x_j$ and $!y_i \& y_j$ 
  * $x_i < x_j$ and $y_i \& !y_j$
  * $x_i \& !x_j$ and $y_i < y_j$
  * $!x_i \& x_j$ and $y_i > y_j$
  * $x_i \& !x_j$ and $!y_i \& y_j$
  * $!x_i \& x_j$ and $y_i \& !y_j$

### Considering Ties

Tied values do not contribute to eeither of the concordant or discordant pair counts, and the original Kendall-tau formula which calculates the tau-a statistic does not consider the presence of tied values.
However, the related tau-b statistic does handle the presence of tied values by adding the tied $x$ and $y$ values to the denominator, and in our special case of missing data, we can add the ties that result from both of $(x_i, x_j)$ and $(y_i, y_j)$ being missing [@kendall_treatment_ties_1945; @kendall_rankcorrelationbook_1948].

$$\tau = \frac{\left | pairs_{concordant} \right | - \left | pairs_{discordant} \right |}{\sqrt{(n_0 - n_{xtie})(n_0 - n_{ytie})}}$$

In the first instance, we remove those x-y points where both values are missing.
We refer to this case as the "local" ICI-Kt correlation.
It is most appropriate for the comparison of only two samples, where we are concerned with what values are present in the two samples, with the odd case of missingness.

The other case, where we leave ties resulting from points with missing $x$ and $y$, we refer to as the "global" ICI-Kt correlation.
In this case, every single sample-sample correlation will consider the same number of pair comparisons.
This is more useful when analyzing and interpreting correlations across multiple samples, not just two feature vectors.

### Theoretical Maxima

The "global" case also provides an interesting property, whereby we can calculate the theoretical maximum correlation that would be possible to observe given the lowest number of shared missing values.
This value can be useful to scale the rest of the observed correlation values across many sample - sample correlations, providing a value that scales an entire dataset appropriately.

$$cor_{max} = \frac{\binom{n-m}{2} + n * m}{\binom{n-m}{2} + n * m + \binom{m}{2}}$$
To calculate the theoretical maximum correlation for rescaling, we choose the two samples with the least number of missing values across the dataset.
Then $n$ is the length of the variables, $m$ is the count of missing values across the two samples divided by two and rounded towards zero.

### Implementation Details

We produced an initial reference implementation in R [@rcoreteam_rlanguage_2020] where the various concordant and discordant pair definitions were written as simple logical tests in R to allow further exploration and validation of faster implementations.
In practice, it is possible to replace the missing values with a value that is smaller than all of the values in the two sets of values under consideration, and then do tests of signs to define concordant and discordant pairs (see the definitions of concordant and discordant pairs above).

After this initial implementation, we used the SciPy [@virtanen_scipy_2020] kendalltau code as a model to create a mergesort based implementation that has a complexity of O(nlog(n)), in comparison to the pairwise testing that has a complexity of O(n^2) [@knight_mergesortkendall_1966].
This provides a 460X speedup of the runtime to compare two feature vectors.
In comparison to directly counting the concordant and discordant pairs, some of the values in the mergesort implementation can reach the 32 bit default limit in C++.
Therefore, we used 64 bit ... **I'm not sure how to word this here**.

C++ (via *Rcpp* [@eddelbuettel_rcppseamless_2011; @eddelbuettel_seamlessbook_2013; @eddelbuettel_extendingrcpp_2018]) and R code implementations are in the src/kendallc.cpp and R/kendalltau.R files of the ICIKendallTau R package, hosted at https://github.com/MoseleyBioinformaticsLab/ICIKendallTau.
The version of the code used in this manuscript is available on figshare.

### Simulated Data Sets

Simulated feature vectors (analytical samples) are generated by drawing 1000 random values from a log-normal distribution with a mean of 1 and sd of 0.5 and sorting them.
A negative analytical sample has values multiplied by -1.
Missing value indices are generated by randomly sampling up to 499 of the lowest values in each sample.
The missing indices are replaced with NA, and then correlations between the analytical samples are calculated.

Another, more realistic, simulated data set is generated by drawing 1000 random values from a log-normal distribution, and adding noise from a normal distribution with a mean of zero and standard deviation of 0.2 to create two statistical samples.
Missing values are created in these statistical samples via two methods: 1) by setting various intensity cutoffs from 0 to 1.5 in 0.1 increments and setting any values below the cutoff to missing; 2) randomly sampling indices in the two-sample matrix in increments of zero to 300 in increments of 50 and setting the indices to missing.

### EGFR RNA-Seq Data Set

RNA-Seq dataset from null and knock-in EGFR mice mutants. **going to need more than this**.

### Adenocarcinoma Recount RNA-Seq Data Set

We downloaded the V2 Recount lung cancer data, extracted the scaled counts, and trimmed to the Stage I adenocarcinoma samples, and those genes that have a non-zero count in at least one of the samples.

Pearson and Kendall correlations are calculated with zero, and replacing zero values with NA.

### Multiprocessing

When a large number of samples need to be compared, it may be useful to split the comparisons across compute instances (across hyperthreaded cores, physical cores, or physical compute nodes).
The *furrr* R package makes the definition of compute clusters easy.
For a large correlation matrix computation, we first define the sample - sample comparisons to be performed, then check how many instances of compute are available (defined by a previous call to *furrr*), and finally split the comparisons into a list that can be easily distributed using the *furrr::future_map* function.

### Subsampling of Features

In addition to using the full set of features to calculate sample - sample correlations, it is possible to select a subsample of features and only use this feature subset to calculate correlations.
Three methods were implemented: 1) random selections; 2) top set of features by variance; and 3) top set of features by variance of principal components.

For (1), random selection, feature subsamples of 1 - 10% by 1% increments, and then in 5% increments to 90% were taken using the *sample* R function.
For (2), selection by variance, feature variances across samples were calculated, and then ranked by the variances.
For (3), selection by variance of the principal components, principal component analysis (PCA) decomposition was performed on the full data set, the variance contribution of each PC calculated, and then the percent variance for each principal component (PC) is used to determine the number of features with high loadings to select from the given PC.
In this feature subsampling approach, weaker PCs contribute proportionately fewer features to the feature subset.
Both the feature variance and PC variance subsample methods were run with subsamples of 30% and 50% of the features.

In all subsamples, the sample - sample correlation differences to those generated with the full feature set were compared, and the median and standard deviation of all of the correlation differences calculated to evaluate the performance of the subsampling.



### Computing Environment

Most calculations were run on a virtual machine with 80 virtual cores, and 1 TB of RAM.
The virtual machine is running on top of a 50 node cluster, each with 4 10-core processors, 3TB of RAM and an 8TB solid-state-drive array, provided by the Kentucky Research Informatics Cloud (KyRIC).
KyRIC manages the virtual machines via an OpenStack instance.
We used the *drake* package to manage calculation dependencies [@landau_drakepackage_2018].
For the comparisons of time taken using different numbers of samples to evaluate the algorithmic complexity, calculations were run on a single laptop Intel i7 core clocked at 2.2 GHz.

## Results


### Comparison To Other Correlation Measures

We compared the ICI-Kt correlation to both Pearson and regular Kendall-tau-b correlations as calculated by the built-in R functions using simulated data sets with missing values, as well as a large RNA-seq dataset.

We created two samples with 1000 observations each drawn from a log-normal distribution, and sorted in each case.
The *true* correlation for each of the Kendall and Pearson correlations were calculated, and then missingness was introduced in the lower range of values (see Methods).

```{r compare_realistic}
loadd(realistic_sample_1)
loadd(realistic_sample_2)
loadd(realistic_neg_sample)
ref_pearson = cor(realistic_sample_1, realistic_sample_2)
ref_kendall = ici_kendalltau(rbind(realistic_sample_1, realistic_sample_2), exclude_0 = FALSE, scale_max = FALSE)$cor[2,1]

ref_pearson_neg = cor(realistic_sample_1, realistic_neg_sample)
ref_kendall_neg = ici_kendalltau(rbind(realistic_sample_1, realistic_neg_sample), exclude_0 = FALSE, scale_max = FALSE)$cor[2,1]

loadd(realistic_na)
n_na = purrr::map_int(realistic_na, length)
loadd(realistic_positive_pearson)
loadd(realistic_positive_kendall)
loadd(realistic_positive_kt)
loadd(realistic_negative_pearson)
loadd(realistic_negative_kendall)
loadd(realistic_negative_kt)
positive_df = rbind(data.frame(cor = realistic_positive_pearson,
                                diff = ref_pearson - realistic_positive_pearson,
                                na = n_na,
                                type = "Pearson",
                                dir = "positive"),
                     data.frame(cor = realistic_positive_kendall,
                                diff = ref_kendall - realistic_positive_kendall,
                                na = n_na,
                                type = "Kendall",
                                dir = "positive"),
                     data.frame(cor = realistic_positive_kt,
                                diff = ref_kendall - realistic_positive_kt,
                                na = n_na,
                                type = "ICI-Kt",
                                dir = "positive"))
negative_df = rbind(data.frame(cor = realistic_negative_pearson,
                                diff = ref_pearson_neg - realistic_negative_pearson,
                                na = n_na,
                                type = "Pearson",
                                dir = "negative"),
                     data.frame(cor = realistic_negative_kendall,
                                diff = ref_kendall_neg - realistic_negative_kendall,
                                na = n_na,
                                type = "Kendall",
                                dir = "negative"),
                     data.frame(cor = realistic_negative_kt,
                                diff = ref_kendall_neg - realistic_negative_kt,
                                na = n_na,
                                type = "ICI-Kt",
                                dir = "negative"))
negative_df = negative_df %>%
  dplyr::mutate(perc_missing = na / max(na) * 100,
                type = factor(type, levels = c("Kendall", "Pearson", "ICI-Kt"), ordered = TRUE))
positive_df = positive_df %>%
  dplyr::mutate(perc_missing = na / max(na) * 100,
                type = factor(type, levels = c("Kendall", "Pearson", "ICI-Kt"), ordered = TRUE))
pos_diff = ggplot(positive_df, aes(x = perc_missing, y = diff)) + 
  geom_point() +
  facet_grid(type ~ ., scales = "free") +
  labs(subtitle = "Positive Correlation",
       x = "Percent Missing") +
  scale_y_continuous(labels = label_scientific(digits = 2))
neg_diff = ggplot(negative_df, aes(x = perc_missing, y = diff)) +
  geom_point() +
  facet_grid(type ~ ., scales = "free") +
  labs(subtitle = "Negative Correlation",
       x = "Percent Missing") +
  scale_y_continuous(labels = label_scientific(digits = 2))
pos_diff + neg_diff
```

**Fig X**. The effect on Kendall, Pearson and ICI-Kt correlations as increasing percentages in either or both samples are replaced with missing values for both positively and negatively correlated samples.

In **Figure X**, we can see that as missing values are added, only the ICI-Kt correlation values change in any significant way.
As the number of missing values increase, the ICI-Kt values drop or increase by up to 0.2.
Similarly, Pearson correlation is also affected, but the degree of change in the correlation values are much less (notice the difference in the y-axis scales), on the order of only 1x10-4 for the positive case.

### Utility in Large Omics' Data Sets

#### Detecting Outlier Samples

We analyzed two RNA-Seq datasets for possible outliers using ICI-Kt.
The first is a collection of 56 samples in a cancer model and trying to determine differences in expression with and without EGFR as well as grown flat or as organoids.
ICI-Kt correlations were calculated using the genes that had a non-zero count in at least one of the samples.
For each sample, the median ICI-Kt value was calculated amongst the samples for each of the defined groups.
Two samples were found to have low correlations within their group.


```{r outlier_adenocarcinoma_setup}
loadd(ref_cor)
loadd(transcript_data)
adeno_cor = ref_cor$cor
adeno_info = readRDS(here::here("data", "recount_adeno_info.rds"))
adeno_cor = adeno_cor[adeno_info$sample_id2, adeno_info$sample_id2]

transcript_data = transcript_data[, adeno_info$sample_id2]
pearson_adeno_cor = cor(log1p(transcript_data))

med_cor = median_correlations(adeno_cor, sample_classes = adeno_info$sample_type)
p_med_cor = median_correlations(pearson_adeno_cor, sample_classes = adeno_info$sample_type)

split_med = split(med_cor, med_cor$sample_class)
split_p = split(p_med_cor, p_med_cor$sample_class)
cor_outlier = function(cor_df){
  out_vals = boxplot.stats(cor_df$med_cor)$out
  cor_df$sample_id[cor_df$med_cor %in% out_vals]
}
med_outliers = purrr::map(split_med, cor_outlier) %>% unlist()
med_cor$outlier = FALSE
med_cor[med_cor$sample_id %in% med_outliers, "outlier"] = TRUE
p_outliers = purrr::map(split_p, cor_outlier) %>% unlist()
p_med_cor$outlier = FALSE
p_med_cor[p_med_cor$sample_id %in% p_outliers, "outlier"] = TRUE
p_med_cor$correlation = "Pearson"
med_cor$correlation = "ICI-Kt"

all_med_cor = rbind(p_med_cor, med_cor)
all_med_cor = all_med_cor %>%
  dplyr::mutate(type2 = dplyr::case_when(
    grepl("Primary", sample_class) ~ "Tumor",
    grepl("Normal", sample_class) ~ "Normal"
  ))
```

The second is a collection of `r ncol(transcript_data)` adenocarcinoma samples from the TCGA project, and incorporated into the recount2 project [@collado-torres_recount2_2017].
In this dataset there are `r nrow(dplyr::filter(adeno_info, grepl("Normal", sample_type)))` normal and `r nrow(dplyr::filter(adeno_info, grepl("Tumor", sample_type)))` tumor samples.

From the pairwise sample-sample correlations, median correlations for each sample with all the other samples of the same class (normal or tumor) were calculated.
Outliers were determined using the boxplot.stats R function.

```{r plot_adenocarcinoma}
ggplot(all_med_cor, aes(x = type2, y = med_cor)) + 
  geom_sina(aes(color = outlier, group = sample_class)) +
  facet_wrap(~ correlation) +
  theme(legend.position = c(0.4, 0.2)) +
  labs(x = "Sample Type", y = "Median Correlation")
```

For the adenocarcinoma data, median correlations within the tumor and normal groups were calculated.
`r tools::toTitleCase(n2w(nrow(dplyr::filter(med_cor, outlier, grepl("Normal", sample_class)))))` sample in the Normal, and `r n2w(nrow(dplyr::filter(med_cor, outlier, grepl("Tumor", sample_class))))` outliers were found, whether using ICI-Kt correlation or Pearson correlation.

### Algorithmic Performance

We compared the performance of our Rcpp mergesort implementation to the base R cor function using both "pearson" and "kendall" methods on a single core with increasing numbers of features.

Figure X 

```{r show_performance_graph}
single_core_perf = run_single_cor()
s_perf_plot = ggplot(single_core_perf, aes(x = n, y = time, color = method)) +
  geom_line(size = 2) + 
  coord_cartesian(ylim = c(NA, 0.002)) + 
  theme(legend.position = c(0.6, 0.2)) +
  labs(x = "Number of Features", y = "Time (s)")
#s_perf_plot

mod_perf = dplyr::filter(single_core_perf, method %in% "ici_kt")
mod_perf$time = mod_perf$time + 0.015
core_perf2 = bind_rows(
  mod_perf,
  dplyr::filter(single_core_perf, !(method %in% "ici_kt"))
)
inset_perf = ggplot(core_perf2, aes(x = n, y = time, color = method)) +
  geom_line(size = 1.5) +
  theme(legend.position = "none",
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank()) +
  labs(x = "", y = "")

s_perf_plot + inset_element(inset_perf, 0.1, 0.72, 1, 1)
```

Figure X. Comparison of the Rcpp mergesort ici_kt, and base R's correlation function using the "pearson" and "kendall" methods as the number of features are increased from 100 to 5000.
Inset is the expanded view showing the full range of computation time in seconds.


### Multi-Processing Performance

In addition to an evaluation of the mergesort performance, we also wanted to evaluate the multi-processing performance.
For this, we ran with small numbers of samples **without** parallelization enabled, and then with larger numbers of samples **with** parallelization enabled.

```{r dependence_number}
loadd(combined_big)
combined_big = combined_big %>%
  dplyr::mutate(set = "large")

ggplot(combined_big, aes(x = n_sample, y = log2(time))) + 
  geom_line() +
  facet_wrap(~ set, nrow = 1, scales = "free_x") +
  labs(x = "Number of Samples",
       y = "Log2(Time) (s)")
```

**Fig X**. Time taken by ICI-Kt as a function of the number of samples with multi-processing at the R level, using 80 cores.

As samples are added, the run time quickly increases logarithmically, similar to increasing samples on a single core.


### Feature Subsampling Randomly

```{r pretty_print}
pretty_print = function(in_val){
  sprintf(in_val, fmt = '%#.1f')
}
```

```{r load_data}
loadd(transcript_data)
loadd(ref_cor)
```

```{r load_ici_timings, fig.height = 8, fig.width = 5}
loadd(run_random_select_random_fraction_0.01)
loadd(run_random_select_random_fraction_0.1)
loadd(run_random_select_random_fraction_0.3)
diff_vals = data.frame(ref = extract_data(ref_cor),
                       f_01 = 
extract_data(run_random_select_random_fraction_0.01),
                     f_1 = extract_data(run_random_select_random_fraction_0.1),
                     f_3 = extract_data(run_random_select_random_fraction_0.3))
diff_vals = diff_vals %>%
  dplyr::mutate(diff_1 = f_01 - ref,
                diff_10 = f_1 - ref,
                diff_30 = f_3 - ref)
ref_f1_f3 = 
  ((ggplot(diff_vals, aes(x = ref, y = f_01)) + geom_point() +
     labs(x = "All Features", y = "1% Features") +
  coord_equal() +
    geom_abline(slope = 1, color = "red")) |
  (ggplot(diff_vals, aes(x = diff_1)) +
     geom_histogram(bins = 100) +
     labs(x = "Correlation Residual") +
     coord_cartesian(xlim = c(-0.03, 0.03)) +
     scale_y_continuous(expand = expansion(mult = 0)) +
     geom_vline(xintercept = 0, color = "red"))) /
  ((ggplot(diff_vals, aes(x = ref, y = f_1)) + geom_point() +
     labs(x = "All Features", y = "10% Features") +
     coord_equal() +
      geom_abline(slope = 1, color = "red")) | 
  (ggplot(diff_vals, aes(x = diff_10)) +
     geom_histogram(bins = 100) +
     labs(x = "Correlation Residual") +
     coord_cartesian(xlim = c(-0.03, 0.03)) +
     scale_y_continuous(expand = expansion(mult = 0)) +
     geom_vline(xintercept = 0, color = "red"))) /
  ((ggplot(diff_vals, aes(x = ref, y = f_3)) + geom_point() +
     labs(x = "All Features", y = "30% Features") +
     coord_equal() +
      geom_abline(slope = 1, color = "red")) +
  (ggplot(diff_vals, aes(x = diff_30)) +
     geom_histogram(bins = 100) +
     labs(x = "Correlation Residual") +
     coord_cartesian(xlim = c(-0.03, 0.03)) +
     scale_y_continuous(expand = expansion(mult = 0)) +
     geom_vline(xintercept = 0, color = "red")))

ref_f1_f3
```

Figure X. Sample-sample ICI-Kt correlations using all features compared with 1%, 10% and 30% random subsets of features. Left: The sample-sample correlations using all features are plotted against the sample-sample correlations using a random subset. Red line indicates perfect agreement. Right: Histogram of the residual differences of the sample-sample correlations, where the difference is the (random subset - all features).

Given the length of time it takes to run the calculations for larger datasets (in particular RNA-Seq data), we wondered if "good enough" results could be obtained using subsamples of the features across the samples.
Our starting data is a set of stage I (I, Ia, and Ib) adenocarcinoma tumor and normal tissue RNA-Seq samples from the *recount2* project, with `r nrow(transcript_data)` features that have a non-zero value in at least one sample, and `r ncol(transcript_data)` samples.

In our original implementation with actual pairwise comparisons, the adenocarcinoma pairwise comparisons took 6.7 hours, on an 80 core virtual machine.
With the mergesort implementation, that time was greatly reduced.
However, given that there may still be cases where the number of comparisons are so large that run time is still a concern, we investigated the effect of estimating sample-sample correlations using a subset of the features.

Figure X compares the sample-sample correlations from a random percent subsample of features to those obtained using all of the features.
Both the pairwise plot (left) and histogram of residual differences (right) show that the correlations become much closer to those obtained when using all of the features.


### Discussion

???

### Conclusions

???

## Author Contributions

RMF wrote the code in the visualizationQualityControl R package, tested the ICI-Kt correlation code, and wrote all of the analysis code for this manuscript. HNBM conceived of the ICI-Kt correlation, provided input into code structures, provided supervision of the analyses and interpretation of results. Both authors contributed to the writing of the manuscript.

## Supplemental

The simplest simulated data set includes three samples, where two are perfectly correlated, and the third perfectly anti-correlated. 
We then created missing values (replaced value with NA) in each sample systematically, and computed the ICI-Kt, Pearson, and Kendall-tau correlations.

```{r compare_simple}
loadd(positive_kt)
loadd(positive_pearson)
positive_pearson[is.na(positive_pearson)] = 0

compare_df = data.frame(ici_kt = positive_kt,
                        pearson = positive_pearson)
ggplot(compare_df, aes(x = pearson, y = ici_kt)) + geom_point()
```

In addition to comparing their performance, we can check that the algorithmic complexity fits the theoretically expected complexity by fitting a regression line of the run time to the number of items.

```{r plot_complexity}
complex_figure = create_complexity_figure(single_core_perf)
complex_figure
```

## Acknowledgements

The results shown here are in whole or part based upon data generated by the TCGA Research Network: https://www.cancer.gov/tcga.

## References
