---
title: 'Supplemental Materials'
author: 'Robert M Flight, Praneeth S Bhatt, and Hunter NB Moseley'
date: '`r Sys.time()`'
output: 
  rmarkdown::word_document
bibliography: '`r here::here("doc/icikt_references.json")`'
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning = FALSE, 
                      message = FALSE, fig.width = 8, 
                      fig.height = 6, fig.keep = "all",
                      tab.cap.style = "paragraph",
                      tab.cap.pre = "",
                      tab.cap.sep = "")

increment_count = function(in_object, id_name){
  n_obj = length(in_object)
  use_number = max(in_object) + 1
  in_object = c(in_object, use_number)
  names(in_object)[n_obj + 1] = id_name
  in_object
}

paste_label = function(pre_text, in_object, id_name){
  object_number = in_object[id_name]
  
  if (length(object_number) > 1) {
    object_numbers = paste(object_number, collapse = ", ")
  } else {
    object_numbers = object_number
  }
  
  use_text = paste(pre_text, object_numbers, sep = "")
  
  use_text
}

figure_count = c("_" = 0)
table_count = c("_" = 0)
```

## Gierlinski Yeast Replicate Outliers

Gierlinski et al [@gierlinski_statisticalmodels_2015] proposed a combination of median sample-sample correlation, gene count outlier fraction, and a chi-square test of read coverage per gene to score each biological replicate from a group of samples.
Outside of defining this combined score, they do not describe the actual criteria of saying a replicate is "bad".
In this work, we used **only** the sample-sample median correlation to identify "bad" or "outlier" samples, using the boxplot.stats function to determine outliers, where an outlier is defined as samples that are > 1.5x away from the limits of the box defining the distribution.

Here, we calculated Pearson correlation using the raw counts, and with only those genes that had non-zero reads in both samples.
This version of Pearson correlation recreates the median sample-sample correlations observed in the original report.

Finally, we note the actual samples recorded as outliers by Gierlinski and coworkers, noted as "Manuscript".

```{r load_yeast_outliers}
figure_count = increment_count(figure_count, "yeast_outliers")

loadd(yeast_paper_outliers)

all_yeast = purrr::map_df(c("yeast_outliers_1",
                            "yeast_outliers_25",
                            "yeast_outliers_50"), function(.x){
                              tmp = readd(.x, character_only = TRUE)
                              tmp$outliers
                            }) %>%
  perc_to_number()

all_yeast = add_method(all_yeast)

yeast_single = all_yeast %>%
  dplyr::select(-method) %>%
  dplyr::filter(keep_num == "1")
tmp_yeast = yeast_single %>%
  dplyr::filter(which %in% "pearson_base_nozero")
tmp_yeast$outlier = FALSE
tmp_yeast[tmp_yeast$sample_id %in% yeast_paper_outliers, "outlier"] = TRUE
tmp_yeast$which = "manuscript"
yeast_single2 = rbind(yeast_single, tmp_yeast)

compare_yeast = c("icikt", "icikt_complete", "pearson_log", "pearson_base_nozero", "manuscript")
```

```{r yeast_single_plot}
yeast_single2 %>%
  add_method(map_method = c("icikt" = "ICI-Kt",
                                                 "icikt_complete" = "ICI-Kt * Completeness",
                                                 "pearson_base" = "Pearson Base",
                                                 "pearson_base_nozero" = "Pearson No Zeros",
                                                 "pearson_log1p" = "Pearson Log(x + 1)",
                                                 "pearson_log" = "Pearson Log(x)",
                                                 "kt_base" = "Kendall-tau",
                                                 "manuscript" = "Manuscript")) %>%
  dplyr::filter(which %in% compare_yeast) %>%
  ggplot(aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) +
  geom_sina() +
  facet_wrap(~ method, nrow = 1)
```

`r paste_label("Figure S", figure_count, "yeast_outliers")`.
Median correlations for each of the yeast RNA-Seq sample to all other samples in the same group, using either ICI-Kt, ICI-Kt * completeness, Pearson correlation, or from the Gierlinski manuscript.


```{r yeast_outlier_table}
yeast_table = compare_outlier_tables(yeast_single2, compare_yeast, sort_var = "manuscript", map_method = c("icikt" = "ICI-Kt",
                                                 "icikt_complete" = "ICI-Kt * Completeness",
                                                 "pearson_base" = "Pearson Base",
                                                 "pearson_base_nozero" = "Pearson No Zeros",
                                                 "pearson_log1p" = "Pearson Log(x + 1)",
                                                 "pearson_log" = "Pearson Log(x)",
                                                 "kt_base" = "Kendall-tau",
                                                 "manuscript" = "Manuscript"))

table_count = increment_count(table_count, "yeast_outliers")
```

`r paste_label("Table S", table_count, "yeast_outliers")`.
Yeast dataset median correlation values and outlier determination for each outlier from each of the correlation methods.

```{r yeast_table2}
#yeast_ft_out = set_caption(yeast_table,
  # caption = paste0(paste_label("Table S", table_count, "yeast_outliers"), ". Yeast dataset median correlation values and outlier determination for each outlier from each of the correlation methods."), style = "paragraph")
yeast_ft_out = autofit(yeast_table)
yeast_ft_out
```

In `r paste_label("Figure S", figure_count, "yeast_outliers")` and `r paste_label("Table S", table_count, "yeast_outliers")` we can see how the determination of outliers was not made solely on the basis of correlation alone, but on a combination of factors that lead to some of the higher correlating samples (using raw counts and Pearson correlaion) being considered outliers where lower correlating samples were not listed as being outliers.

Regardless, using ICI-Kt or ICI-Kt * Completeness in this instance, the outliers using the simple distribution summary statistics and an outlier having to be > 1.5 median error, the outliers are mostly a superset of the outliers determined by Gierlinski et al, with the exception of three samples specific to their data: Snf2.24, WT.22, and WT.28.

It should be noted that it seems that Gierlinski et al used an eyeball cutoff for the outliers based on the combined score of correlation, outlier fraction, and read coverage chi-square fitness, with samples having a combined log-score > -2.8 (evaluated by RMF eyeball on a zoomed in graph and a ruler), a value that is never stated in the manuscript, and which seems arbitrary from the data.

## Brainson RNASeq Samples

```{r brainsonrnaseq_load}
figure_count = increment_count(figure_count, "brainson_outliers")
all_brainson = purrr::map_df(c("brainsonrnaseq_outliers_1",
                            "brainsonrnaseq_outliers_25",
                            "brainsonrnaseq_outliers_50",
                            "brainsonrnaseq_outliers_75",
                            "brainsonrnaseq_outliers_100"), function(.x){
                              tmp = readd(.x, character_only = TRUE)
                              tmp$outliers
                            }) %>%
  perc_to_number()

brainson_single = all_brainson %>%
  dplyr::filter(keep_num == "1")
compare_brainson = c("icikt", "icikt_complete", "pearson_log")

brainson_single %>%
  add_method() %>%
  dplyr::filter(which %in% compare_brainson) %>%
  ggplot(aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) +
  geom_sina() + 
  facet_wrap(~ method, nrow = 1)
```

`r paste_label("Figure S", figure_count, "brainson_outliers")`. 
Brainson RNA-seq sample outliers calculated from median correlations using ICI-Kt only, ICI-Kt with completeness, and Pearson using logged values.

`r paste_label("Figure S", figure_count, "brainson_outliers")` shows the effect of including the pairwise completeness measure, where the "total" tumor samples have a higher correlation than the "sorted" samples where a specific subset of cells were collected from each sample and sequenced.
This is due to having more genes with non-zero reads in general over the sorted samples.

```{r brainson_table}
brainson_table = compare_outlier_tables(brainson_single, compare_brainson, sort_var = "pearson_log")

table_count = increment_count(table_count, "brainson_outliers")
```

`r paste_label("Table S", table_count, "brainson_outliers")`.
Brainson dataset median correlation values and outlier determination for each outlier from each of the correlation methods.

```{r brainson_table2}
# brainson_ft_out = set_caption(brainson_table,
#   caption = paste0(paste_label("Table S", table_count, "brainson_outliers"), ". Brainson dataset median correlation values and outlier determination for each outlier from each of the correlation methods."))
brainson_ft_out = autofit(brainson_table)
brainson_ft_out
```

`r paste_label("Table S", table_count, "brainson_outliers")` indicates the outliers determined with each correlation method.
Discussions with the Brainson group have determined that some of the samples in the sorted group were run on different days with different technicians, which may explain the two sorted outliers.

## TCGA Adenocarcinoma

```{r adeno_load}
figure_count = increment_count(figure_count, "adeno_outliers")
all_adeno = purrr::map_df(c("adeno_outliers_1",
                            "adeno_outliers_25",
                            "adeno_outliers_50"), function(.x){
                              tmp = readd(.x, character_only = TRUE)
                              tmp$outliers
                            }) %>%
  perc_to_number()

new_adeno = unique(all_adeno[, c("sample_id", "sample_class")]) %>%
  dplyr::mutate(sample_seq = seq(1, n()),
                sample_new = dplyr::case_when(
    sample_class %in% "Tumor" ~ paste0("T.", sample_seq),
    TRUE ~ paste0("N.", sample_seq)
  )) %>%
  dplyr::mutate(old_sample = sample_id,
                sample_seq = NULL,
                sample_class = NULL,
                sample_id = sample_new,
                sample_new = NULL)

all_adeno2 = dplyr::left_join(new_adeno, all_adeno, by = c("old_sample" = "sample_id"))

adeno_single = all_adeno2 %>%
  dplyr::filter(keep_num == "1")
compare_adeno = c("icikt", "icikt_complete", "pearson_log")

adeno_single %>%
  add_method() %>%
  dplyr::filter(which %in% compare_adeno) %>%
  ggplot(aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) +
  geom_sina() + 
  facet_wrap(~ method, nrow = 1)
```

`r paste_label("Figure S", figure_count, "adeno_outliers")`.
TCGA Adenocarcinoma outliers using various measures of correlation.

```{r adeno_table}
adeno_table = compare_outlier_tables(adeno_single, compare_adeno, sort_var = "pearson_log")

table_count = increment_count(table_count, "adeno_outliers")
```

`r paste_label("Table S", table_count, "adeno_outliers")`.
TCGA adenocarcinoma dataset median correlation values and outlier determination for each outlier from each of the correlation methods.

```{r adeno_table2}
# adeno_ft_out = set_caption(adeno_table,
#   caption = paste0(paste_label("Table S", table_count, "adeno_outliers"), ". adeno dataset median correlation values and outlier determination for each outlier from each of the correlation methods."))
adeno_ft_out = autofit(adeno_table)
adeno_ft_out
```

```{r summarize_adeno}
table_count = increment_count(table_count, "adeno_mad")
adeno_variation = adeno_single %>%
  add_method() %>%
  dplyr::filter(which %in% compare_adeno) %>%
  calculate_variation() %>%
  dplyr::select(mad) 

adeno_ratios = adeno_variation %>%
  tidyr::pivot_wider(names_from = sample_class, values_from = mad) %>%
  dplyr::summarise(ratio = format(Tumor / Normal, digits = 0)) %>%
  tidyr::pivot_wider(names_from = method, values_from = ratio)
adeno_variation_wide = adeno_variation %>% 
  tidyr::pivot_wider(names_from = sample_class, values_from = mad) %>%
  flextable() %>%
  colformat_double(digits = 3) %>%
  autofit() 
```

`r paste_label("Table S", table_count, "adeno_mad")`.
Adenocarcinoma median-absolute-deviations (MAD) for each group of sample median correlations in each type of tissue.

```{r summarize_adeno2}
# %>%
#   set_caption(caption = paste0(paste_label("Table S", table_count, "adeno_mad"), ". Adenocarcinoma median-absolute-deviations (MAD) for each group of sample median correlations in each type of tissue."), style = "paragraph")
adeno_variation_wide
```

As `r paste_label("Figure S", figure_count, "adeno_outliers")` and `r paste_label("Table S", table_count, "adeno_outliers")`, although there are not many outliers in these relatively large groups, Pearson correlation using log-tansformed values results in the most outliers.
Also of note is that ICI-Kt * Completeness is the only measure where the variance measure for both groups is similar.
For both ICI-Kt and Pearson correlation, the tumor samples variance measures are `r adeno_ratios[["ICI-Kt"]]`x and `r adeno_ratios[["Pearson Log(x)"]]`x greater than normal, respectively.


## Effect of Increasing Presence in Samples

One sure way to mess with the correlation values is to impose the condition that a gene has a non-zero count in a minimum percentage of the samples in one of the classes.
In all of the example plots above, and the tables in the main manuscript text, a gene had to be present in at least **one sample**.
Here, we explore what happens if we increase the number of samples required for a gene to be present in at 25% and 50% of a class for the yeast samples and adenocarcinoma samples, and 25%, 50%, 75% and all samples of a class for the Brainson samples.
We also examine the effect using all possible correlation measures.

### Yeast Samples

```{r yeast_by_method}
figure_count = increment_count(figure_count, "yeast_by_method")
ggplot(all_yeast, aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) + 
  geom_sina() +
  facet_grid(keep_num ~ method)
```

`r paste_label("Figure S", figure_count, "yeast_by_method")`.
Median correlations by correlation method and applying different fractional cutoffs.

```{r yeast_by_fraction}
figure_count = increment_count(figure_count, "yeast_by_keep_num")
ggplot(all_yeast, aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) +
  geom_sina() +
  facet_grid(method ~ keep_num)
```

`r paste_label("Figure S", figure_count, "yeast_by_keep_num")`.
Median correlations by applying different fractional cutoffs and different correlation methods.


```{r yeast_plot_differences}
figure_count = increment_count(figure_count, "yeast_differences")
yeast_summary = calculate_variation(all_yeast)
yeast_diffs = calculate_differences(yeast_summary)

ggplot(yeast_diffs, aes(x = keep_num, y = diff)) + 
  geom_col() +
  facet_grid(which_diff ~ method) +
  labs(x = "Keep Fraction", y = "WT / Snf2 Differences")
```

`r paste_label("Figure S", figure_count, "yeast_differences")`.
Difference between WT and Snf2 group medians and MAD for the different correlation methods and gene presence fractions.


### Brainson Samples

```{r brainson_plot1}
figure_count = increment_count(figure_count, "brainson_by_method")

brainson_classes = c("sorted", "total")
all_brainson_method = add_method(all_brainson) %>%
  dplyr::filter(sample_class %in% brainson_classes)
ggplot(all_brainson_method, aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) + 
  geom_sina() +
  facet_grid(keep_num ~ method)
```

`r paste_label("Figure S", figure_count, "brainson_by_method")`.
Median correlations by correlation method and applying different fractional cutoffs for Brainson RNA-seq data.


```{r brainson_plot2}
figure_count = increment_count(figure_count, "brainson_by_keep_num")
ggplot(all_brainson_method, aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) + 
  geom_sina() +
  facet_grid(method ~ keep_num)
```

`r paste_label("Figure S", figure_count, "brainson_by_keep_num")`.
Median correlations by correlation method and applying different fractional cutoffs for Brainson RNA-seq data.


```{r brainson_plot_differences}
figure_count = increment_count(figure_count, "brainson_differences")
brainson_summary = calculate_variation(all_brainson_method)
brainson_diffs = calculate_differences(brainson_summary)

ggplot(brainson_diffs, aes(x = keep_num, y = diff)) + 
  geom_col() +
  facet_grid(which_diff ~ method) +
  labs(x = "Keep Fraction", y = "Sorted / Total Differences")
```

`r paste_label("Figure S", figure_count, "brainson_differences")`.
Difference between sorted and total sample group medians and MAD for the different correlation methods and gene presence fractions.


### Adenocarcinoma

```{r adeno_by_method}
adeno_method = add_method(all_adeno2)
figure_count = increment_count(figure_count, "adeno_by_method")
ggplot(adeno_method, aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) + 
  geom_sina() +
  facet_grid(keep_num ~ method)
```

`r paste_label("Figure S", figure_count, "adeno_by_method")`. 
Median correlations by correlation method and applying different fractional cutoffs for TCGA adenocarcinoma RNA-seq data.

```{r adeno_by_fraction}
figure_count = increment_count(figure_count, "adeno_by_keep_num")
ggplot(adeno_method, aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) +
  geom_sina() +
  facet_grid(method ~ keep_num)
```

`r paste_label("Figure S", figure_count, "adeno_by_keep_num")`. 
Median correlations by correlation method and applying different fractional cutoffs for TCGA adenocarcinoma RNA-seq data.

```{r adeno_plot_differences}
figure_count = increment_count(figure_count, "adeno_differences")
adeno_summary = calculate_variation(adeno_method)
adeno_diffs = calculate_differences(adeno_summary)

ggplot(adeno_diffs, aes(x = keep_num, y = diff)) + 
  geom_col() +
  facet_grid(which_diff ~ method) +
  labs(x = "Keep Fraction", y = "Normal / Tumor Differences")
```

`r paste_label("Figure S", figure_count, "adeno_differences")`. 
Difference of normal and tumor medians and MADs  by correlation method and applying different fractional cutoffs for TCGA adenocarcinoma RNA-seq data.
