---
title: 'Supplemental Materials'
author: 'Robert M Flight, Praneeth S Bhatt, and Hunter NB Moseley'
date: '`r Sys.time()`'
output: 
  rmarkdown::word_document:
    keep_md: true
bibliography: '`r here::here("doc/icikt_references.json")`'
csl: plos-computational-biology.csl
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
targets::tar_source(files = c("./packages.R", "R"))

unlink(here::here("doc/supplemental_materials_files"), recursive = TRUE)

supp_figure_count = dn_counter$new("Figure ", "_", "S")
supp_table_count = dn_counter$new("Table ", "_", "S")


knitr::opts_chunk$set(echo = FALSE, warning = FALSE, 
                      message = FALSE, fig.width = 8, 
                      fig.height = 6, fig.keep = "all",
                      fig.process = supp_figure_count$modify_path,
                      tab.cap.style = "paragraph",
                      tab.cap.pre = "",
                      tab.cap.sep = "",
                      dpi = 600,
                      dev = "ragg_png")



```

```{r}
#| label: load
tar_load(c(lod_graph_adenocarcinoma,
           lod_graph_egfrgenotype,
           lod_graph_egfrgenotypetumorculture,
           lod_graph_nsclc,
           lod_graph_ratstamina,
           positive_kt,
           positive_pearson,
           negative_kt,
           all_kt,
           negative_pearson,
           censored_value_plots,
           complexity_figure,
           yeast_paper_outliers,
           yeast_outliers,
           yeast_single,
           egfrgenotype_outliers,
           egfrgenotype_single,
           egfrgenotypetumorculture_outliers,
           egfrgenotypetumorculture_single,
           adenocarcinoma_outliers,
           adenocarcinoma_single,
           nsclc_outliers,
           nsclc_single,
           ratstamina_outliers,
           ratstamina_single))
```

```{r}
#| label: increment
supp_figure_count$increment(c("lod_adenocarcinoma",
                              "lod_egfrgenotype",
                              "lod_egfrgenotypetumorculture",
                              "lod_nsclc",
                              "lod_ratstamina",
                            "kt_pearson",
                            "ici_distribution",
                            "kt_distribution",
                            "leftcensored",
                            "complexity",
                            "yeast_outliers",
                            "egfrgenotype_outliers",
                            "egfrgenotypetumorculture_outliers",
                            "adenocarcinoma_outliers",
                            "nsclc_outliers",
                            "ratstamina_outliers",
                            "adenocarcinoma_by_method",
                            "adenocarcinoma_by_keep_num",
                            "adenocarcinoma_differences",
                            "yeast_by_method",
                            "yeast_by_keep_num",
                            "yeast_differences",
                            "egfrgenotype_by_method",
                            "egfrgenotype_by_keep_num",
                            "egfrgenotype_differences"))



supp_table_count$increment(c("yeast_outliers",
                           "egfrgenotype_outliers",
                           "adenocarcinoma_outliers",
                           "adenocarcinoma_mad",
                           "nsclc_outliers",
                           "ratstamina_outliers"))
```

## Limits of Detection Issues in Datasets

In addition to the yeast RNA-seq dataset, we investigated the relationship of median abundance and minimum median abundance of present features with the number of present features across samples for the other four datasets, including TCGA Adenocarcinoma RNA-seq, Brainson Tumor Cultures RNA-seq, RCSIRM Non small cell lung cancer lipidomics, and rat diabetes metabolomics from Metabolomics Workbench.

As shown in `r supp_figure_count$label_text(c("lod_adenocarcinoma", "lod_tumorcultures", "lod_nsclc", "lod_ratstamina"))`, there is a direct relationship between the minimum median value and number of samples a feature is present in.

```{r}
#| label: lod_adenocarcinoma 
#| fig.width: 10
#| fig.height: 8
wrap_plots(lod_graph_adenocarcinoma, ncol = 1) + plot_annotation(tag_levels = "A")
```
`r supp_figure_count$label_text("lod_adenocarcinoma")`.
Adenocarcinoma dataset count medians by number of samples feature was present in.
**A**: Log10 of medians for present values copmared to the number of samples the feature was present in.
**B**: Log10 of minimum median for each value of number of samples the feature was present in.

```{r}
#| label: lod_egfrgenotype
#| fig.width: 10
#| fig.height: 8 
wrap_plots(lod_graph_egfrgenotype, ncol = 1) + plot_annotation(tag_levels = "A")
```
`r supp_figure_count$label_text("lod_egfrgenotype")`.
EGFR genotype dataset count medians by number of samples feature was present in.
**A**: Log10 of medians for present values copmared to the number of samples the feature was present in.
**B**: Log10 of minimum median for each value of number of samples the feature was present in.

```{r}
#| label: lod_egfrgenotypetumorculture
#| fig.width: 10
#| fig.height: 8 
wrap_plots(lod_graph_egfrgenotypetumorculture, ncol = 1) + plot_annotation(tag_levels = "A")
```
`r supp_figure_count$label_text("lod_egfrgenotypetumorculture")`.
EGFR genotype and tumor culture dataset count medians by number of samples feature was present in.
**A**: Log10 of medians for present values copmared to the number of samples the feature was present in.
**B**: Log10 of minimum median for each value of number of samples the feature was present in.

```{r}
#| label: lod_nsclc
#| fig.width: 10
#| fig.height: 8
wrap_plots(lod_graph_nsclc, ncol = 1) + plot_annotation(tag_levels = "A")
```
`r supp_figure_count$label_text("lod_nsclc")`.
Non-small-cell-lung-cancer (NSCLC) lipidomics dataset count medians by number of samples feature was present in.
**A**: Log10 of medians for present values copmared to the number of samples the feature was present in.
**B**: Log10 of minimum median for each value of number of samples the feature was present in.


```{r}
#| label: lod_ratstamina
#| fig.width: 10
#| fig.height: 8
wrap_plots(lod_graph_ratstamina, ncol = 1) + plot_annotation(tag_levels = "A")
```
`r supp_figure_count$label_text("lod_ratstamina")`.
Rat stamina dataset count medians by number of samples feature was present in.
**A**: Log10 of medians for present values copmared to the number of samples the feature was present in.
**B**: Log10 of minimum median for each value of number of samples the feature was present in.

## Simulated Data

### Simple Data Set

The simplest simulated data set includes three samples, where two are perfectly correlated, and the third perfectly anti-correlated. 
We then created missing values (replaced value with NA) in each sample systematically, and computed the ICI-Kt, Pearson, and Kendall-tau correlations.

```{r kt_pearson, dn_id = supp_figure_count}
all_kt_other = rbind(positive_kt, negative_kt)
all_pearson = rbind(positive_pearson, negative_pearson)
all_pearson[is.na(all_pearson$cor), "cor"] = 0

compare_df = data.frame(ici_kt = all_kt_other$cor,
                        pearson = all_pearson$cor)
set.seed(1234)
rand_rows = sample(nrow(compare_df), 10000)
compare_df2 = compare_df[rand_rows, ]
ggplot(compare_df2, aes(x = pearson, y = ici_kt)) + 
  geom_point() +
  labs(x = "Pearson", y = "ICI-Kt")
```

`r supp_figure_count$label_text("kt_pearson")`. Comparison of ICI-Kt and Pearson correlations for perfectly positive and negatively correlated samples, systematically replacing values with NA.
NA values from Pearson were replaced with zero for this comparison.

We can also examine the full set of positive and negative correlations generated as we vary the number of missing entries between two positively correlated samples and two negatively correlated samples.
These distributions are shown in `r supp_figure_count$label_text("kt_distribution")`.
We can see that the distributions from both ICI-Kt and Kendall-tau are the same, which is expected given we replaced missing values (NA) with zero *within* the ICI-Kt code, and replaced missing values (NA) with zero prior to calculating Kendall-tau correlations.

```{r ici_distribution, dn_id = supp_figure_count}
ici_plot = ggplot(all_kt, aes(x = ici_kt)) +
  geom_histogram(bins = 100) +
  facet_wrap(~ comp, ncol = 1) +
  labs(x = "ICI-Kt Correlation", y = "count")
ici_plot
```

`r supp_figure_count$label_text("ici_distribution")`.
ICI-Kendall-tau correlation as missing values are varied between two samples.

```{r kt_distribution, dn_id = supp_figure_count}
kt_plot = ggplot(all_kt, aes(x = kendall)) +
  geom_histogram(bins = 100) +
  facet_wrap(~ comp, ncol = 1) +
  labs(x = "Kendall-tau Correlation", y = "count")
kt_plot
```

`r supp_figure_count$label_text("kt_distribution")`.
Kendall-tau correlation as missing values are varied between two samples and replaced with 0 before calculating Kendall-tau.

### Semi-Realistic Data Set

```{r leftcensored, fig.height = 8, fig.width = 9, dn_id = supp_figure_count}
wrap_plots(censored_value_plots, ncol = 1) + plot_annotation(tag_levels = "A")
```

`r supp_figure_count$label_text("leftcensored")`.
Effect of introducing missing values from a cutoff (A & B) or randomly (C) on different measures of correlation, including ICI-Kt, Kendall with pairwise complete, Kendall replacing missing with 0, Pearson with pairwise complete, and Pearson replacing missing with 0. 
A) Missing values introduced by setting an increasing cutoff. 
B) Missing values introduced by setting an increasing cutoff, and then log-transforming the values before calculating correlation. 
C) Missing values introduced at random. For the random case, each sample of random positions was repeated 100 times.

## Algorithmic Complexity

In addition to comparing their performance, we can check that the algorithmic complexity fits the theoretically expected complexity by fitting a regression line of the run time to the number of items.
The run times and fitted lines for each method (R's Pearson, the ICI-Kt, and R's Kendall-tau) are shown in `r supp_figure_count$label_text("complexity")`.

```{r complexity, dn_id = supp_figure_count}
complexity_figure
```

`r supp_figure_count$label_text("complexity")`. 
Time in seconds needed as a function of the number of features, with a fitted line for the assumed complexity for each of the methods tested, including R's Pearson correlation, the ICI-Kt mergesort, and R's Kendall-tau correlation algorithm.


## Outlier Samples

### Gierlinski Yeast

Gierliński et al [@gierlinski_statisticalmodels_2015] proposed a combination of median sample-sample correlation, gene count outlier fraction, and a chi-square test of read coverage per gene to score each biological replicate from a group of samples.
Outside of defining this combined score, they do not describe the actual criteria of saying a replicate is "bad".
In this work, we used **only** the sample-sample median correlation to identify "bad" or "outlier" samples, using the boxplot.stats function to determine outliers, where an outlier is defined as samples that are > 1.5x away from the limits of the box defining the distribution.

Here, we calculated Pearson correlation using the raw counts, and with only those genes that had non-zero reads in both samples.
This version of Pearson correlation recreates the median sample-sample correlations observed in the original report.

Finally, we note the actual samples recorded as outliers by Gierlinski and coworkers, noted as "Manuscript".

In `r supp_figure_count$label_text("yeast_outliers")` and `r supp_table_count$label_text("yeast_outliers")` we can see how the determination of outliers was not made solely on the basis of correlation alone, but on a combination of factors that lead to some of the higher correlating samples (using raw counts and Pearson correlaion) being considered outliers where lower correlating samples were not listed as being outliers.

Regardless, using ICI-Kt or ICI-Kt * Completeness in this instance, the outliers using the simple distribution summary statistics and an outlier having to be > 1.5 median error, the outliers are mostly a superset of the outliers determined by Gierlinski et al, with the exception of three samples specific to their data: Snf2.24, WT.22, and WT.28.

It should be noted that it seems that Gierlinski et al used an eyeball cutoff for the outliers based on the combined score of correlation, outlier fraction, and read coverage chi-square fitness, with samples having a combined log-score > -2.8 (evaluated by RMF eyeball on a zoomed in graph and a ruler), a value that is never stated in the manuscript, and which seems arbitrary from the data.

```{r load_yeast_outliers}
manual_method = c("icikt" = "IK",
                                                 "icikt_complete" = "IKC",
                                                 "pearson_base" = "PB",
                                                 "pearson_base_nozero" = "PN0",
                                                 "pearson_log1p" = "PL1",
                                                 "pearson_log" = "PL",
                                                 "kt_base" = "Kt",
                                                 "manuscript" = "M")

other_method = c("icikt" = "IK",
                                                 "icikt_complete" = "IKC",
                                                 "pearson_base" = "PB",
                                                 "pearson_base_nozero" = "PN0",
                                                 "pearson_log1p" = "PL1",
                                                 "pearson_log" = "PL",
                                                 "kt_base" = "Kt")



all_yeast = rbind_outliers(yeast_outliers) %>%
  perc_to_number()

all_yeast = add_method(all_yeast)

yeast_single = yeast_single %>%
  perc_to_number()
tmp_yeast = yeast_single %>%
  dplyr::filter(which %in% "pearson_base_nozero")
tmp_yeast$outlier = FALSE
tmp_yeast[tmp_yeast$sample_id %in% yeast_paper_outliers, "outlier"] = TRUE
tmp_yeast$which = "manuscript"
yeast_single2 = rbind(yeast_single, tmp_yeast)

compare_yeast = c("icikt", "icikt_complete", "pearson_base", "pearson_base_nozero", "pearson_log1p", "pearson_log", "kt_base", "manuscript")
```

```{r yeast_outliers, dn_id = supp_figure_count}
yeast_single2 %>%
  add_method(map_method = manual_method) %>%
  dplyr::filter(which %in% compare_yeast) %>%
  ggplot(aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) +
  geom_sina() +
  facet_wrap(~ method, nrow = 1) +
  labs(x = "Sample Class", y = "Median Correlation")
```

`r supp_figure_count$label_text("yeast_outliers")`.
Median correlations for each of the yeast RNA-Seq sample to all other samples in the same group, using different correlation measures on different data. Abbreviations for different measures and data are: IK: ICI-Kt; IKC: ICI-Kt * Completeness; M: Manuscript; PL1: Pearson Log(x + 1); PN0: Pearson No Zeros.


```{r yeast_outlier_table}
yeast_table = compare_outlier_tables(yeast_single2, compare_yeast, sort_var = "manuscript", map_method = manual_method)
```



### EGFR RNASeq Genotype

`r supp_table_count$label_text("egfrgenotype_outliers")` indicates the outliers determined with each correlation method.
Discussions with the Brainson group have determined that some of the samples in the sorted group were run on different days with different technicians, which may explain the two sorted outliers.

`r supp_figure_count$label_text("egfrgenotype_outliers")` visualizes the sample-specific median of each metric calculated from the Brainson RNA-Seq data set.
Likewise, `r supp_table_count$label_text("egfrgenotype_outliers")` lists the outliers detected based on each metric.
In this example, Pearson shows superior sensitivity to outlier detection, likely due to the larger spread in median Pearson correlation values.

```{r egfrgenotype_load}

all_egfrgenotype = rbind_outliers(egfrgenotype_outliers) %>%
  perc_to_number()

compare_egfrgenotype = c("icikt", "icikt_complete", "pearson_base", "pearson_base_nozero", "pearson_log1p", "pearson_log", "kt_base")
```

```{r egfrgenotype_outliers, dn_id = supp_figure_count}
egfrgenotype_single %>%
  add_method(other_method) %>%
  dplyr::filter(which %in% compare_egfrgenotype) %>%
  ggplot(aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) +
  geom_sina() +
  facet_wrap(~ method, nrow = 1) +
  labs(x = "Sample Class", y = "Median Correlation") +
  theme(axis.text.x = element_text(angle = 90))
```

`r supp_figure_count$label_text("egfrgenotype_outliers")`.
Median correlations for each of the Brainson RNA-Seq sample to all other samples in the same group, using different correlation measures on different data. 
Abbreviations for different measures and data are: IK: ICI-Kt; IKC: ICI-Kt * Completeness; PB: Pearson Base, PN: Pearson No Zero; PL1: Pearson Log(x + 1); PL: Pearson Log(x); Kt: Kendall-tau.


`r supp_figure_count$label_text("egfrgenotype_outliers")` shows the effect of including the pairwise completeness measure, where the "total" tumor samples have a higher correlation than the "sorted" samples where a specific subset of cells were collected from each sample and sequenced.
This is due to having more genes with non-zero reads in general over the sorted samples.

```{r egfrgenotype_table}
egfrgenotype_table = compare_outlier_tables(egfrgenotype_single, compare_egfrgenotype, "pearson_log1p", map_method = other_method)
```


### EGFR RNASeq Genotype & Tumor Type

```{r}
#| label: setup_egfrgenotypetumorculture
all_egfrgenotypetumorculture_outliers = rbind_outliers(egfrgenotypetumorculture_outliers) |>
  perc_to_number()
```

```{r}
#| label: egfrgenotypetumorculture_outliers
egfrgenotypetumorculture_single %>%
  add_method(other_method) %>%
  dplyr::filter(which %in% compare_egfrgenotype) %>%
  ggplot(aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) +
  geom_sina() +
  facet_wrap(~ method, nrow = 1) +
  labs(x = "Sample Class", y = "Median Correlation") +
  theme(axis.text.x = element_text(angle = 90))
```


```{r}
#| label: egfrgenotypetumorculture_table
egfrgenotypetumorculture_table = compare_outlier_tables(egfrgenotypetumorculture_single, compare_egfrgenotype, "pearson_log1p", map_method = other_method)
```

### TCGA Adenocarcinoma

```{r adenocarcinoma_load}
all_adeno = rbind_outliers(adenocarcinoma_outliers) %>%
  perc_to_number()

new_adeno = unique(all_adeno[, c("sample_id", "sample_class")]) %>%
  dplyr::mutate(sample_seq = seq(1, n()),
                sample_new = dplyr::case_when(
    sample_class %in% "Tumor" ~ paste0("T.", sample_seq),
    TRUE ~ paste0("N.", sample_seq)
  )) %>%
  dplyr::mutate(old_sample = sample_id,
                sample_seq = NULL,
                sample_class = NULL,
                sample_id = sample_new,
                sample_new = NULL)

all_adeno2 = dplyr::left_join(new_adeno, all_adeno, by = c("old_sample" = "sample_id"))
adenocarcinoma_single = dplyr::left_join(adenocarcinoma_single, new_adeno, by = c("sample_id" = "old_sample"))

compare_adeno = c("icikt", "icikt_complete", "pearson_base", "pearson_base_nozero", "pearson_log1p", "pearson_log", "kt_base")
```

```{r adenocarcinoma_outliers, dn_id = supp_figure_count}
adenocarcinoma_single %>%
  add_method(map_method = other_method) %>%
  dplyr::filter(which %in% compare_adeno) %>%
  ggplot(aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) +
  geom_sina() +
  facet_wrap(~ method, nrow = 1) +
  labs(x = "Sample Class", y = "Median Correlation") +
  theme(axis.text.x = element_text(angle = 90))
```

`r supp_figure_count$label_text("adenocarcinoma_outliers")`.
Median correlations for each of the TGCA adenocarcinoma RNA-Seq sample to all other samples in the same group, using different correlation measures on different data. 
Abbreviations for different measures and data are: IK: ICI-Kt; IKC: ICI-Kt * Completeness; PB: Pearson Base, PN: Pearson No Zero; PL1: Pearson Log(x + 1); PL: Pearson Log(x); Kt: Kendall-tau.


```{r adenocarcinoma_table}
adenocarcinoma_table = compare_outlier_tables(adenocarcinoma_single, compare_adeno, "pearson_log1p", map_method = other_method)
```

```{r summarize_adeno}
adenocarcinoma_variation = adenocarcinoma_single %>%
  add_method(map_method = other_method) %>%
  dplyr::filter(which %in% compare_adeno) %>%
  calculate_variation() %>%
  dplyr::select(mad) 

adenocarcinoma_ratios = adenocarcinoma_variation %>%
  tidyr::pivot_wider(names_from = sample_class, values_from = mad) %>%
  dplyr::summarise(ratio = format(Tumor / Normal, digits = 1, nsmall = 0)) %>%
  tidyr::pivot_wider(names_from = method, values_from = ratio)
adenocarcinoma_variation_wide = adenocarcinoma_variation %>% 
  tidyr::pivot_wider(names_from = sample_class, values_from = mad) %>%
  flextable() %>%
  colformat_double(digits = 3) %>%
  autofit() 
```

`r supp_table_count$label_text("adenocarcinoma_mad")`.
Adenocarcinoma median-absolute-deviations (MAD) for each group of sample median correlations in each type of tissue.


```{r summarize_adeno2}
# %>%
#   set_caption(caption = paste0(paste_label("Table S", supp_table_count, "adenocarcinoma_mad"), ". Adenocarcinoma median-absolute-deviations (MAD) for each group of sample median correlations in each type of tissue."), style = "paragraph")
autofit(adenocarcinoma_variation_wide)
```

As `r supp_figure_count$label_text("adenocarcinoma_outliers")` and `r supp_table_count$label_text("adenocarcinoma_outliers")`, although there are not many outliers in these relatively large groups, Pearson correlation using log-transformed values results in the most outliers.
Also of note is that ICI-Kt * Completeness is the only measure where the variance measure for both groups is similar.
For both ICI-Kt and Pearson correlation, the tumor samples variance measures are `r adenocarcinoma_ratios[["ICI-Kt"]]`x and `r adenocarcinoma_ratios[["Pearson Log(x + 1)"]]`x greater than normal, respectively.

`r supp_figure_count$label_text("adenocarcinoma_outliers")` visualizes the sample-specific median metric value calculated from the TCGA RNA-Seq data set, with `r supp_table_count$label_text("adenocarcinoma_outliers")` listing the outliers detected from each metric.
In this example, the ICI-Kt * completeness correlation demonstrated superior sensitivity for outlier detection, likely due to its better sensitivity to measurement completeness.
The wider spread of measurement completeness is expected, since TCGA is an aggregate of samples collected from various human cancer patents as compared to the first two examples which were collected from specific experiments in cell culture (`r supp_table_count$label_text("adenocarcinoma_mad")`).
The wider spread of measurement completeness is expected, since TCGA is an aggregate of samples collected from various human cancer patients as compared to the first two examples which were collected from specific experiments in cell culture. 
However, the outliers detected are not a complete superset of the outliers detected by ICI-Kt and Pearson, probably due to the integration of completeness and correlation.


### NSCLC Lipidomics

```{r}
#| label: setup_nsclc
all_nsclc_outliers = rbind_outliers(nsclc_outliers) |>
  perc_to_number()
```

```{r}
#| label: nsclc_outliers
nsclc_single %>%
  add_method(other_method) %>%
  dplyr::filter(which %in% compare_egfrgenotype) %>%
  ggplot(aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) +
  geom_sina() +
  facet_wrap(~ method, nrow = 1) +
  labs(x = "Sample Class", y = "Median Correlation") +
  theme(axis.text.x = element_text(angle = 90))
```


```{r}
#| label: nsclc_table
nsclc_table = compare_outlier_tables(nsclc_single, compare_egfrgenotype, "pearson_log1p", map_method = other_method)
```


### Rat Stamina Metabolomics

```{r}
#| label: setup_ratstamina
all_ratstamina_outliers = rbind_outliers(ratstamina_outliers) |>
  perc_to_number()
```

```{r}
#| label: ratstamina_outliers
ratstamina_single %>%
  add_method(other_method) %>%
  dplyr::filter(which %in% compare_egfrgenotype) %>%
  ggplot(aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) +
  geom_sina() +
  facet_wrap(~ method, nrow = 1) +
  labs(x = "Sample Class", y = "Median Correlation") +
  theme(axis.text.x = element_text(angle = 90))
```


```{r}
#| label: ratstamina_table
ratstamina_table = compare_outlier_tables(ratstamina_single, compare_egfrgenotype, "pearson_log1p", map_method = other_method)
```


## Effect of Increasing Presence in Samples

One sure way to limit the effect of missing values on correlation values is to impose the condition that a gene has a non-zero count in a minimum percentage of the samples in one of the classes.
In all of the example plots above, and the tables in Supplemental Tables, a gene had to be present in at least **one sample**.
Here, we explore what happens if we increase the number of samples required for a gene to be present in at 25% and 50% of a class for the yeast samples and adenocarcinoma samples, and 25%, 50%, 75% and all samples of a class for the Brainson samples.


### Yeast Samples


```{r yeast_by_method, dn_id = supp_figure_count}
all_yeast2 = all_yeast %>%
  add_method(other_method) %>%
  clear_outliers()
ggplot(all_yeast2, aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) + 
  geom_sina() +
  facet_grid(keep_num ~ method) +
  labs(x = "Sample Class", y = "Median Correlation")
```

`r supp_figure_count$label_text("yeast_by_method")`.
Median correlations by correlation method and applying different fractional cutoffs.
Abbreviations for different measures and data are: IK: ICI-Kt; IKC: ICI-Kt * Completeness; Kt: Kendall-tau; PB: Pearson Base (raw values); PL: Pearson Log(x); PL1: Pearson Log(x + 1); PN0: Pearson No Zeros.

```{r yeast_by_keep_num, dn_id = supp_figure_count}
ggplot(all_yeast2, aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) +
  geom_sina() +
  facet_grid(method ~ keep_num) +
  labs(x = "Sample Class", y = "Median Correlation")
```

`r supp_figure_count$label_text("yeast_by_keep_num")`.
Median correlations by applying different fractional cutoffs and different correlation methods.
Abbreviations for different measures and data are: IK: ICI-Kt; IKC: ICI-Kt * Completeness; Kt: Kendall-tau; PB: Pearson Base (raw values); PL: Pearson Log(x); PL1: Pearson Log(x + 1); PN0: Pearson No Zeros.

```{r yeast_differences, dn_id = supp_figure_count}
yeast_summary = calculate_variation(all_yeast2)
yeast_diffs = calculate_differences(yeast_summary)

ggplot(yeast_diffs, aes(x = keep_num, y = diff)) + 
  geom_col() +
  facet_grid(which_diff ~ method) +
  labs(x = "Keep Fraction", y = "WT / Snf2 Differences") +
  theme(axis.text.x = element_text(angle = 90))
```

`r supp_figure_count$label_text("yeast_differences")`.
Difference between WT and Snf2 group medians and MAD for the different correlation methods and gene presence fractions.
Abbreviations for different measures and data are: IK: ICI-Kt; IKC: ICI-Kt * Completeness; Kt: Kendall-tau; PB: Pearson Base (raw values); PL: Pearson Log(x); PL1: Pearson Log(x + 1); PN0: Pearson No Zeros.

### EGFR Genotype Samples

```{r increment_count_brainson}

```

```{r egfrgenotype_by_method, dn_id = supp_figure_count}
egfrgenotype_classes = c("wt", "het", "null")
all_egfrgenotype_method = add_method(all_egfrgenotype, map_method = other_method) %>%
  dplyr::filter(sample_class %in% egfrgenotype_classes)
ggplot(all_egfrgenotype_method, aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) + 
  geom_sina() +
  facet_grid(keep_num ~ method) +
  theme(axis.text.x = element_text(size = 10)) +
  labs(x = "Sample Class", y = "Median Correlation")
```

`r supp_figure_count$label_text("egfrgenotype_by_method")`.
Median correlations by correlation method and applying different fractional cutoffs for Brainson RNA-seq data.
Abbreviations for different measures and data are: IK: ICI-Kt; IKC: ICI-Kt * Completeness; Kt: Kendall-tau; PB: Pearson Base (raw values); PL: Pearson Log(x); PL1: Pearson Log(x + 1); PN0: Pearson No Zeros.

```{r egfrgenotype_by_keep_num, dn_id = supp_figure_count}
ggplot(all_egfrgenotype_method, aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) + 
  geom_sina() +
  facet_grid(method ~ keep_num) +
  labs(x = "Sample Class", y = "Median Correlation")
```

`r supp_figure_count$label_text("egfrgenotype_by_keep_num")`.
Median correlations by correlation method and applying different fractional cutoffs for Brainson RNA-seq data.
Abbreviations for different measures and data are: IK: ICI-Kt; IKC: ICI-Kt * Completeness; Kt: Kendall-tau; PB: Pearson Base (raw values); PL: Pearson Log(x); PL1: Pearson Log(x + 1); PN0: Pearson No Zeros.

```{r egfrgenotype_differences, dn_id = supp_figure_count}
egfrgenotype_summary = calculate_variation(all_egfrgenotype_method)
egfrgenotype_diffs = calculate_differences(egfrgenotype_summary)

ggplot(egfrgenotype_diffs, aes(x = keep_num, y = diff)) + 
  geom_col() +
  facet_grid(which_diff ~ method) +
  labs(x = "Keep Fraction", y = "Sorted / Total Differences") +
  theme(axis.text.x = element_text(angle = 90))
```

`r supp_figure_count$label_text("egfrgenotype_differences")`.
Difference between sorted and total sample group medians and MAD for the different correlation methods and gene presence fractions.
Abbreviations for different measures and data are: IK: ICI-Kt; IKC: ICI-Kt * Completeness; Kt: Kendall-tau; PB: Pearson Base (raw values); PL: Pearson Log(x); PL1: Pearson Log(x + 1); PN0: Pearson No Zeros.

### Adenocarcinoma


```{r adenocarcinoma_by_method, dn_id = supp_figure_count}
adenocarcinoma_method = add_method(all_adeno2, other_method) %>%
  clear_outliers()
ggplot(adenocarcinoma_method, aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) + 
  geom_sina() +
  facet_grid(keep_num ~ method) +
  theme(axis.text.x = element_text(angle = 90)) +
  labs(x = "Sample Class", y = "Median Correlation")
```

`r supp_figure_count$label_text("adenocarcinoma_by_method")`. 
Median correlations by correlation method and applying different fractional cutoffs for TCGA adenocarcinoma RNA-seq data.
Abbreviations for different measures and data are: IK: ICI-Kt; IKC: ICI-Kt * Completeness; Kt: Kendall-tau; PB: Pearson Base (raw values); PL: Pearson Log(x); PL1: Pearson Log(x + 1); PN0: Pearson No Zeros.

```{r adenocarcinoma_by_keep_num, dn_id = supp_figure_count}
ggplot(adenocarcinoma_method, aes(x = sample_class, y = med_cor, color = outlier, group = sample_class)) +
  geom_sina() +
  facet_grid(method ~ keep_num) +
  labs(x = "Sample Class", y = "Median Correlation")
```

`r supp_figure_count$label_text("adenocarcinoma_by_keep_num")`. 
Median correlations by correlation method and applying different fractional cutoffs for TCGA adenocarcinoma RNA-seq data.
Abbreviations for different measures and data are: IK: ICI-Kt; IKC: ICI-Kt * Completeness; Kt: Kendall-tau; PB: Pearson Base (raw values); PL: Pearson Log(x); PL1: Pearson Log(x + 1); PN0: Pearson No Zeros.

```{r adenocarcinoma_differences, dn_id = supp_figure_count}
adenocarcinoma_summary = calculate_variation(adenocarcinoma_method)
adenocarcinoma_diffs = calculate_differences(adenocarcinoma_summary)

ggplot(adenocarcinoma_diffs, aes(x = keep_num, y = diff)) + 
  geom_col() +
  facet_grid(which_diff ~ method) +
  labs(x = "Keep Fraction", y = "Normal / Tumor Differences") +
  theme(axis.text.x = element_text(angle = 90))
```

`r supp_figure_count$label_text("adenocarcinoma_differences")`. 
Difference of normal and tumor medians and MADs  by correlation method and applying different fractional cutoffs for TCGA adenocarcinoma RNA-seq data.
Abbreviations for different measures and data are: IK: ICI-Kt; IKC: ICI-Kt * Completeness; Kt: Kendall-tau; PB: Pearson Base (raw values); PL: Pearson Log(x); PL1: Pearson Log(x + 1); PN0: Pearson No Zeros.

```{r save_supp_objects}
saveRDS(supp_figure_count, file = here::here("doc/supp_figure_count.rds"))
saveRDS(supp_table_count, file = here::here("doc/supp_table_count.rds"))
save(yeast_table, egfrgenotype_table, egfrgenotypetumorculture_table, adenocarcinoma_table, nsclc_table, ratstamina_table, file = here::here("doc/supp_stuff.rda"))
```

## References
